api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,265 @@
+#pragma once
+
+#include <atomic>
+
+#include "opentelemetry/context/context_value.h"
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+// The context class provides a context identifier. Is built as a linked list
+// of context nodes which each have access to the nodes before it, but not
+// after. So each time a write is performed a new context is added to the list.
+// Within each context node there is another linked list which holds the key
+// and value data for each node.
+class Context
+{
+
+public:
+  nostd::shared_ptr<Context> *context_ptr_;
+
+  // Creates a new context by calling a constructor with the passed in
+  // keyValueIterable and returns a shared_ptr pointing to it.
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  static nostd::shared_ptr<Context> CreateContext(T &keys_and_values)
+  {
+    nostd::shared_ptr<Context> temp = nostd::shared_ptr<Context>{new Context(keys_and_values)};
+    temp->context_ptr_              = &temp;
```

Comment:
I don't fully understand the use of `context_ptr_`. It seems it is a pointer to a shared pointer of the object itself. How is it different from just accessing `this`?

Reply:
Because I want both the user and the node right behind this node to be using the same shared_ptr.

Reply:
If I just accessed this when creating a shared_ptr to return to the user and a shared_ptr to point to the next node, then wouldn't they not be linked together and attempt to deconstruct the context twice?

Reply:
Consider using [enable_shared_from_this](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this) for this use case. A pointer to a shared pointer looks dangerous.

Reply:
Perhaps it is what is cause my window test failures.

Reply:
Thanks, i'll give it a try, I assume it works smoothly with nostd::shared_ptr?

Reply:
But in that case wouldn't I need to keep a shared_ptr pointing to the object as a member of the object? Doesn't that defeat the purpose?

Reply:
I am on board, will be rewriting to avoid this issue.

Reply:
You can disregard my comment, I think using `enable_shared_from_this` is ABI stable.

I tried to come up with another suggestion [here](https://github.com/open-telemetry/opentelemetry-cpp/pull/180/files#r457619122).